core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSail.java

```diff
@@ -0,0 +1,311 @@
+/*******************************************************************************
+ * Copyright (c) 2016 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+
+package org.eclipse.rdf4j.sail.inferencer.fc;
+
+import org.eclipse.rdf4j.common.iteration.Iterations;
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Resource;
+import org.eclipse.rdf4j.model.Statement;
+import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
+import org.eclipse.rdf4j.repository.Repository;
+import org.eclipse.rdf4j.repository.RepositoryConnection;
+import org.eclipse.rdf4j.repository.RepositoryResult;
+import org.eclipse.rdf4j.sail.NotifyingSail;
+import org.eclipse.rdf4j.sail.SailException;
+import org.eclipse.rdf4j.sail.inferencer.InferencerConnection;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.StampedLock;
+import java.util.stream.Collectors;
+
+/**
+ * @author Håvard Mikkelsen Ottestad
+ */
+
+public class FastRdfsForwardChainingSail extends AbstractForwardChainingInferencer {
+
+    NotifyingSail data;
+    Repository schema;
+
+    static final Random random = new Random();
+
+    private StampedLock readWriteLock = new StampedLock();
```

Comment:
StampedLock is the only lock I found that allows upgrades.

Read and write is a bit of a poor choice of words. Call them concurrent and serial instead. All threads can concurrently read, add, delete and update the Abox, these will only fail if the underlying sail fails. But add, delete and update to the Tbox need to be done serially. 

The write lock is used to lock writes to the Tbox, and only the Tbox. And there is no way to decide upfront if the user is going to add Tbox statements. 

@jeenbroekstra that is a feature of the stamped lock, but it also provides strong read and write locks, which is what I'm using.